### PR TITLE
KNL-1540 write events using prepared statement batching

### PR DESCRIPTION
**Pull request by ottenhoff**
This is only lightly tested on MySQL so far. I found this chunk of code using YourKit and found that it wasn't executing prepared statement batches despite having rewriteBatchedStatements=true on JDBC connection

### CODE REVIEW

File: kernel/kernel-impl/src/main/java/org/sakaiproject/event/impl/ClusterEventTracking.java [hunk 1]
```diff
@@ -401,42 +401,38 @@ protected void writeBatchEvents(Collection<Event> events)
...
-			// Note: investigate batch writing via the jdbc driver: make sure we can still use prepared statements (check out host arrays, too)
-			// -ggolden
+			// common preparation for each insert
+			String statement = insertStatement();
 
-            // common preparation for each insert
-            String statement = insertStatement();
-            Object fields[] = new Object[6];
+			// Setup a batch of events if not using a cluster
+			Set<Object[]> eventSet = new HashSet<>();
```
Comment (ottenhoff) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/event/impl/ClusterEventTracking.java]: This should probably be a LinkedHashSet to maintain order of items, right?

Comment (buckett) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/event/impl/ClusterEventTracking.java]: Why not have the API use a `List` as in most cases order is always important isn't it?

Comment (ottenhoff) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/event/impl/ClusterEventTracking.java]: Yup, order can be important. I will make the change.

File: kernel/kernel-impl/src/main/java/org/sakaiproject/db/impl/BasicSqlService.java [hunk 2]
```diff
@@ -1153,6 +1154,49 @@ public int dbWriteCount(String sql, Object[] fields, String lastField, Connectio
...
+		{
+			LOG.warn("Sql.dbWriteBatch()", e);
+		}
+		catch (SQLException e)
+		{
+			LOG.warn("Sql.dbWriteBatch(): error code: " + e.getErrorCode() + " sql: " + sql + " " + e);
+		}
+		finally
+		{
+			try
```
Comment (buckett) [inline kernel/kernel-impl/src/main/java/org/sakaiproject/db/impl/BasicSqlService.java]: I know this is the existing pattern in this file so changing it makes it stand out but we could use try-with-resources for this instead.

I'm on the fence about this.

Comment (ottenhoff) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/db/impl/BasicSqlService.java]: Yeah I considered but thought maybe we could update the entire file in one swoop instead

Comment (bjones86) [reply kernel/kernel-impl/src/main/java/org/sakaiproject/db/impl/BasicSqlService.java]: :+1: for code cleanup!

**Comment by buckett** [comment]
What's the reason for closing this one?

**Comment by ottenhoff** [comment]
I haven't been able to show any compelling performance advantages in my Yourkit monitoring. I'll continue trying.

**Pull request closed without merging**